lxd/storage/drivers/driver_lvm_volumes.go

```diff
@@ -804,54 +805,59 @@ func (d *lvm) UnmountVolume(vol Volume, keepBlockDev bool, op *operations.Operat
 
 	refCount := vol.MountRefCountDecrement()
 
-	// Check if already mounted.
-	if vol.contentType == ContentTypeFS && filesystem.IsMountPoint(mountPath) {
-		if refCount > 0 {
-			d.logger.Debug("Skipping unmount as in use", logger.Ctx{"volName": vol.name, "refCount": refCount})
-			return false, ErrInUse
+	// For VMs, unmount the filesystem volume.
+	if vol.IsVMBlock() {
+		fsVol := vol.NewVMBlockFilesystemVolume()
+		ourUnmount, err = d.UnmountVolume(fsVol, false, op)
+
+		// If the VMBlockFilesystem volume is still in use, we use the refCount
+		// of the block volume instead.
```

Comment:
Sorry to push back on this, but I'm still not clear on why this will break if the FS volume is mounted on its own. If some future code calls `MountVolume` on just the FS volume, then we expect that code to also call `UnmountVolume`, so it should keep the refCount accurate.

I guess the real question for me here is: why should trying to unmount the block volume return ErrInUse if the FS volume is still mounted elsewhere? This code (to my knowledge) just considers them as separate volumes for the purpose of refCounting.

I'm just not sure I understand the situation that breaks this :slightly_frowning_face:

Reply:
No worries, quite alright.

My worry here is that we are ignoring an InUse error from the underlying VM config filesystem volume and masking it to the caller of `UnmountVolume` of the VM block volume itself, which previously we did not.

Now, because at the current time our expectation is that the ref counters for the VM block volume and the VM config filesystem volume are in lock-step, this shouldnt cause a problem in practice as on line 820 we are then detecting the refCount > 0 of the VM block volume and returning `ErrInUse` anyway.

So I think the comment we discussed adding on line 813 should be more a statement of that assumption and why its ok to ignore the ErrInUse error there.